### ISSUE TITLE
Chp.2 How the backpropagation algorithm works

### ISSUE THREAD

**Issue by hysic** [hysic/NNDL_ReadingNotes#2]
反向传播（backpropagation）：计算cost function关于w和b的偏微分。

要想应用反向传播，cost function需要满足两个条件：

1. The cost function can be written as an average C=1/n*∑C_x over cost functions C_x for individual training examples, x.
2. The cost function can be written as a function of the outputs from the neural network.

满足上述两个条件的cost function举例：

![image](https://cloud.githubusercontent.com/assets/5415646/26623306/e0ad9c22-461f-11e7-9137-215f982cd630.png)

**Comment by hysic** [comment]
定义：
![image](https://cloud.githubusercontent.com/assets/5415646/26660201/e2b1d5e0-46a8-11e7-8013-1b26838080af.png) 与 
![image](https://cloud.githubusercontent.com/assets/5415646/26775604/2ce75ab0-4a08-11e7-9448-eb29720d0f64.png)


反向传播（BP）算法公式：
![image](https://cloud.githubusercontent.com/assets/5415646/26627131/fb44de72-462b-11e7-8b68-c0c1da01bedf.png)

反向传播算法步骤（单个training example）：
![image](https://cloud.githubusercontent.com/assets/5415646/26660377/482f4e88-46aa-11e7-95cd-11e59e659bb0.png)

BP + mini-batch Gradient descent:
![image](https://cloud.githubusercontent.com/assets/5415646/26662418/4dad3224-46b6-11e7-8983-81b7222cc258.png)

**Comment by hysic** [comment]
反向传播（BP）算法的速度：一次反向传播，就可以算出所有的梯度。所需要的时间：一次前向传播 + 一次后向传播 ≈ 两次前向传播。

对比根据梯度定义（见下式）计算所需要的时间：前向传播的次数 = 权重的个数 + 1

![image](https://cloud.githubusercontent.com/assets/5415646/26773676/ddaf2800-49fe-11e7-83b5-8f796bb2701b.png)

所以BP算法比梯度定义的算法快得多。

**Comment by hysic** [comment]
BP算法公式(3)和(4)的推导：

![image](https://cloud.githubusercontent.com/assets/5415646/26775315/f87d4fba-4a06-11e7-9747-0dabc0db7c35.png)
